modules/47_regipol/regiCarbonPrice/postsolve.gms

```diff
@@ -213,72 +218,84 @@ loop(ext_regi$regiEmiMktTarget(ext_regi),
 );
 p47_allTargetsConverged_iter(iteration,ext_regi) = p47_allTargetsConverged(ext_regi);
 
+*** define current target to be solved
+p47_currentConvergence_iter(iteration,ttot,ext_regi) = 0;
+loop(ext_regi$regiEmiMktTarget(ext_regi),
+*** solving targets sequentially, i.e. only apply target convergence algorithm if previous yearly targets were already achieved
+  if(not(p47_allTargetsConverged(ext_regi) eq 1), !!no rescale need if all targets already converged
+*** define current target to be solved
+    loop((ttot)$regiANDperiodEmiMktTarget_47(ttot,ext_regi),
+      p47_currentConvergencePeriod(ext_regi) = ttot.val;
+      break$(p47_targetConverged(ttot,ext_regi) eq 0); !!only run target convergence up to the first year that has not converged
+    );
+    loop((ttot)$(regiANDperiodEmiMktTarget_47(ttot,ext_regi) and (ttot.val gt p47_currentConvergencePeriod(ext_regi))),
+      p47_nextConvergencePeriod(ext_regi) = ttot.val;
+      break;
+    );
+    loop((ttot,ttot2,emiMktExt,target_type_47,emi_type_47)$(pm_emiMktTarget(ttot,ttot2,ext_regi,emiMktExt,target_type_47,emi_type_47) AND (ttot2.val eq p47_currentConvergencePeriod(ext_regi))),
+      p47_currentConvergence_iter(iteration,ttot2,ext_regi) = 1;
+    );
+  );
+);
+
+*** define refrence year for calculating mitigation cost slope.
+loop((ext_regi,ttot)$regiANDperiodEmiMktTarget_47(ttot,ext_regi),
+  if(ord(iteration) eq 1,
+    p47_slopeReferenceIteration(iteration,ttot,ext_regi) = 1;
+  elseif(NOT(p47_currentConvergence_iter(iteration,ttot,ext_regi) eq p47_currentConvergence_iter(iteration-1,ttot,ext_regi))),
```

Comment:
I don't get this - why is it the right question to ask if the target convergence check was different between last and this iteration? if I understand correctly, this gets updated to the current iteration if a) last iteration the check was on and now it is not on, and b) last iteration the check was off and now it is on?

Reply:
This line says that, for a given region, whenever you change the target that you are currently updating in comparison to the previous iteration (`NOT(p47_currentConvergence_iter(iteration,ttot,ext_regi) eq p47_currentConvergence_iter(iteration-1,ttot,ext_regi)`), you need to reset the `p47_slopeReferenceIteration` value (`p47_slopeReferenceIteration(iteration,ttot,ext_regi) = current iteration`). 
Otherwise, you just keep the reference iteration value unchanged: `p47_slopeReferenceIteration(iteration,ttot,ext_regi) = p47_slopeReferenceIteration(iteration-1,ttot,ext_regi);`.

This way, whenever you change your target year focus, you will run an initial tax update using a simplified formula, and only from the next iteration onwards the slope will be used to update the tax values using as reference the closest iteration that used the simplified formula.

Take a look at `/p/projects/ecemf/REMIND/2040_scenarios/v04_2023_06_02/output/04_Nzero_57_RpEUEff_bio7p5_limCC_EU27_2023-06-02_17.25.43` for example.

We have two targets for EU27, one by 2030 and another by 2050 (`cm_emiMktTarget` = `2020.2030.EU27_regi.all.year.netGHG_LULUCFGrassi_intraRegBunker 2.122, 2035.2050.EU27_regi.all.year.netGHG_LULUCFGrassi 0.01`)

If you take a look at `p47_currentConvergence_iter`, you will see that at iterations 1, 6, 9 and 12, our focus to update the taxes changes from 2030 to 2050 to 2030 and back to 2050. 

  |   | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
EU27_regi | 2030 | 1 | 1 | 1 | 1 | 1 |   |   |   | 1 | 1 | 1 |   |  
EU27_regi | 2050 |   |   |   |   |   | 1 | 1 | 1 |   |   |   | 1 | 1

This means, 
- the 2030 target converged at iteration 6, so we switched to try to solve the 2050 in the next one.
- Then, we updated the 2050 taxes for three iterations, but at iteration 9 these updates caused the 2030 to not converge anymore.
- We kept the 2050 as is, and got back on updating the 2030 target to converge from iteration 9, and it finally converged at iteration 12.
- The we got back to the 2050 target and finally from iteration 14 onward both targets converged, and kept converging until the end of the model run.   

Now how the tax update was done:

- When you are at iterations that change target focus - iteration 1 by definition and iterations 6, 9 and 12 in the example above - the tax update is defined by a simple rule based on the current deviation.
- When you are at other iterations, the tax update is defined by the slope between the tax and target changes, calculated from the nearest iteration that had a target change focus.
  - When you are at iterations 2 to 5, the tax update is defined by the slope that you observed from iteration 1.
  - When you are at iterations 7 to 8, the tax update is defined by the slope that you observed from iteration 6.
  - When you are at iterations 10 to 11, the tax update is defined by the slope that you observed from iteration 9.
  - When you are at iteration 13 onward, the tax update is defined by the slope that you observed from iteration 12.

The values of `p47_slopeReferenceIteration` reflect the reference iterations to calculate the slope as I described above:

  |   | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29 | 30 | 31 | 32 | 33 | 34 | 35 | 36 | 37 | 38 | 39 | 40 | 41 | 42 | 43 | 44
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
EU27_regi | 2030 | 1 | 1 | 1 | 1 | 1 | 6 | 6 | 6 | 9 | 9 | 9 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12 | 12
EU27_regi | 2050 | 1 | 1 | 1 | 1 | 1 | 6 | 6 | 6 | 9 | 9 | 9 | 12 | 12 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14 | 14

Reply:
thanks for the extended explanation!
what about adding "!! reset the slope if the target that is being analyzed changes" in the `elseif` line to get people thinking in the right direction? 
I had misinterpreted the parameter as "is the target converged", and not "is this the target the model is currently focusing on"